Approximations for SIN/LOG2/EXP2 passing all tests.

hikettei opened the pull request:
# Description

This is the implementation for the approximation bounty.

Added an approximation implementation for SIN/LOG2/EXP2 -> `xsin(x), xlog2(x), xexp2(x)` in tinygrad/fastmath.py

For a detailed description, please see: https://github.com/tinygrad/tinygrad/pull/5085

- FYI:
    - The approximation implementation algorithms are adapted by SLEEF. https://arxiv.org/abs/2001.09258
    - reproduced the code in Julia for review. https://github.com/tinygrad/tinygrad/pull/5085#issuecomment-2185557902
    - Benchmark: https://github.com/tinygrad/tinygrad/pull/5085#issuecomment-2193801823
# Prerequisite

This branch contains significant changes and is divided into several prerequisite PRs.

- ~~https://github.com/tinygrad/tinygrad/pull/5173#issuecomment-2194565111~~ -> https://github.com/tinygrad/tinygrad/pull/5223 (No SHL/SHR is required)
- https://github.com/tinygrad/tinygrad/pull/5172#issuecomment-2194558217
- ~~https://github.com/tinygrad/tinygrad/pull/5185~~ -> Resolved
- ~~https://github.com/tinygrad/tinygrad/pull/5186~~ -> Resolved at https://github.com/tinygrad/tinygrad/pull/5299

hikettei commented:
hmm i didn't expect the AMD tests to still fail. it might need one or two more commits.

hikettei commented:
I spent the entire day debugging AMD, but it's still not working at all.. (I would appreciate any advice if you have some.)

https://github.com/hikettei/tinygrad/actions/runs/9712467156/job/26807301623

uuuvn commented:
> I spent the entire day debugging AMD, but it's still not working at all.. (I would appreciate any advice if you have some.)
> 
> https://github.com/hikettei/tinygrad/actions/runs/9712467156/job/26807301623

Re-ran same tests that failed here on tinybox:

```
(.venv) uuuvn@tiny13:~/src/tinygrad$ git log -1
commit ff44a77b6d60691fd4cbbe5be5b3da5b4caba3d7 (HEAD -> test/faster-approx-1, hikettei/test/faster-approx-1)
Merge: 46d916a3 6b08cb5e
Author: hikettei <hikettei@gmail.com>
Date:   Sat Jun 29 17:12:01 2024 +0900

    Merge remote-tracking branch 'upstream/master' into test/faster-approx-1
(.venv) uuuvn@tiny13:~/src/tinygrad$ python -m pytest -n=6 test/test_ops.py test/test_dtype.py test/test_dtype_alu.py test/test_linearizer.py test/test_randomness.py test/imported/test_indexing.py test/external/external_test_hcq.py --durations=20
=============================================================================================== test session starts ================================================================================================
platform linux -- Python 3.10.12, pytest-8.2.2, pluggy-1.5.0
rootdir: /home/uuuvn/src/tinygrad
plugins: xdist-3.6.1, hypothesis-6.103.2
6 workers [669 items]
..........ss.....................................................................s.......................................................................s.................................................. [ 30%]
....................................................sssssssss............s.......................................................s.....................................................s.................... [ 60%]
............s.........s.........sss.ssss.........s...............sss....s...ss.ss............s.......F..sss...s......s....s.s..........s...sss..ss......sss...s....s..ssssssss..s....s.ss................... [ 91%]
s...s.s.......s....ss..................s.................                                                                                                                                                    [100%]
===================================================================================================== FAILURES =====================================================================================================
______________________________________________________________________________________ TestDTypeALU.test_float_midcast_int32 _______________________________________________________________________________________
[gw3] linux -- Python 3.10.12 /home/uuuvn/src/tinygrad/.venv/bin/python

self = <test.test_dtype_alu.TestDTypeALU testMethod=test_float_midcast_int32>

    @given(strat.floats(width=32, min_value=0, max_value=10.0) if skip_overflow else ht.float32,
>          strat.floats(width=32, min_value=0, max_value=10.0) if skip_overflow else ht.float32,
           ht.int32, strat.sampled_from(binary_operations), strat.sampled_from(integer_binary_operations))

test/test_dtype_alu.py:177:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/test_dtype_alu.py:180: in test_float_midcast_int32
    def test_float_midcast_int32(self, a, b, c, op1, op2): universal_test_midcast(a, b, c, op1, op2, dtypes.float32, dtypes.int32)
test/test_dtype_alu.py:99: in universal_test_midcast
    np.testing.assert_allclose(tensor_value, numpy_value, rtol=1e-6 if getenv("PTX") else 1e-7)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<function assert_allclose.<locals>.compare at 0x798bbeda4c10>, array([2147483647], dtype=int32), array([-2147483648], dtype=int32))
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'strict': False, ...}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError:
E           Not equal to tolerance rtol=1e-07, atol=0
E
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference among violations: 1
E           Max relative difference among violations: -4.65661287e-10
E            ACTUAL: array([2147483647], dtype=int32)
E            DESIRED: array([-2147483648], dtype=int32)
E           Falsifying example: test_float_midcast_int32(
E               self=<test.test_dtype_alu.TestDTypeALU testMethod=test_float_midcast_int32>,
E               a=0.0,  # or any other generated value
E               b=inf,
E               c=0,
E               op1=add,
E               op2=add,
E           )
E           Explanation:
E               These lines were always and only run by failing examples:
E                   /home/uuuvn/src/tinygrad/.venv/lib/python3.10/site-packages/numpy/_core/arrayprint.py:558
E                   /home/uuuvn/src/tinygrad/.venv/lib/python3.10/site-packages/numpy/_core/arrayprint.py:574
E                   /home/uuuvn/src/tinygrad/.venv/lib/python3.10/site-packages/numpy/_core/arrayprint.py:923
E                   /home/uuuvn/src/tinygrad/.venv/lib/python3.10/site-packages/numpy/testing/_private/utils.py:207
E                   /home/uuuvn/src/tinygrad/.venv/lib/python3.10/site-packages/numpy/testing/_private/utils.py:830
E                   (and 2 more with settings.verbosity >= verbose)

/usr/lib/python3.10/contextlib.py:79: AssertionError
================================================================================================= warnings summary =================================================================================================
test/test_ops.py::TestOps::test_std_one_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:895: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,2,3,1,5)], lambda x: x.std(axis=(0,3)), forward_only=True)

test/test_ops.py::TestOps::test_std_one_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:897: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,2,3,1,5)], lambda x: x.std(axis=(0,3), correction=5), forward_only=True)

test/test_ops.py::TestOps::test_std_one_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:900: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,2,3,1,5)], lambda x: x.std(axis=(0,4), correction=5))

test/test_ops.py::TestOps::test_std_zero_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:890: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,0,3,0,5)], lambda x: x.std(axis=(1,3)))

test/test_ops.py::TestOps::test_std_zero_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:891: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,0,3,0,5)], lambda x: x.std(axis=(1,3), correction=0))

test/test_ops.py::TestOps::test_std_zero_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:892: UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,0,3,0,5)], lambda x: x.std(axis=(1,3), correction=5))

test/test_dtype.py::TestHalfDType::test_casts_to
  /home/uuuvn/src/tinygrad/test/test_dtype.py:55: RuntimeWarning: overflow encountered in cast
    _test_op(lambda: a.cast(target_dtype), target_dtype, list(a.numpy().astype(_to_np_dtype(target_dtype))))

test/test_ops.py::TestOps::test_var_one_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:868: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,2,3,1,5)], lambda x: x.var(axis=(0,3)), forward_only=True)

test/test_ops.py::TestOps::test_var_one_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:870: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,2,3,1,5)], lambda x: x.var(axis=(0,3), correction=5), forward_only=True)

test/test_ops.py::TestOps::test_var_one_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:873: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,2,3,1,5)], lambda x: x.var(axis=(0,4), correction=5), forward_only=True)

test/test_ops.py::TestOps::test_var_zero_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:863: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,0,3,0,5)], lambda x: x.var(axis=(1,3)))

test/test_ops.py::TestOps::test_var_zero_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:864: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,0,3,0,5)], lambda x: x.var(axis=(1,3), correction=0))

test/test_ops.py::TestOps::test_var_zero_in_axis
  /home/uuuvn/src/tinygrad/test/test_ops.py:865: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1807.)
    helper_test_op([(1,0,3,0,5)], lambda x: x.var(axis=(1,3), correction=5))

test/test_dtype_alu.py::TestDTypeALU::test_float16
test/test_dtype_alu.py::TestDTypeALU::test_float16
test/test_dtype_alu.py::TestDTypeALU::test_float16
test/test_dtype_alu.py::TestDTypeALU::test_float16
test/test_dtype_alu.py::TestDTypeALU::test_float64
test/test_dtype_alu.py::TestDTypeALU::test_float32
test/test_dtype_alu.py::TestDTypeALU::test_float32
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:65: RuntimeWarning: invalid value encountered in subtract
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)), np.array([b]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py::TestDTypeALU::test_float16
test/test_dtype_alu.py::TestDTypeALU::test_float64
test/test_dtype_alu.py::TestDTypeALU::test_float32
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:65: RuntimeWarning: invalid value encountered in add
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)), np.array([b]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py: 44 warnings
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:76: RuntimeWarning: invalid value encountered in sqrt
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py::TestDTypeALU::test_float32_unary
test/test_dtype_alu.py::TestDTypeALU::test_float16_unary
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:76: RuntimeWarning: divide by zero encountered in reciprocal
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py: 17 warnings
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:76: RuntimeWarning: invalid value encountered in sin
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py: 47 warnings
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:76: RuntimeWarning: invalid value encountered in log
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py::TestDTypeALU::test_float32_unary
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
test/test_dtype_alu.py::TestDTypeALU::test_float16_unary_nan_masked
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:76: RuntimeWarning: overflow encountered in exp
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py::TestDTypeALU::test_float32_unary
test/test_dtype_alu.py::TestDTypeALU::test_float16_unary
test/test_dtype_alu.py::TestDTypeALU::test_float32_unary_nan_masked
test/test_dtype_alu.py::TestDTypeALU::test_float16_unary_nan_masked
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:76: RuntimeWarning: divide by zero encountered in log
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py::TestDTypeALU::test_float64
test/test_dtype_alu.py::TestDTypeALU::test_float32
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:65: RuntimeWarning: overflow encountered in multiply
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)), np.array([b]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py::TestDTypeALU::test_float64
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:65: RuntimeWarning: overflow encountered in add
    numpy_value = op[1](np.array([a]).astype(_to_np_dtype(dtype)), np.array([b]).astype(_to_np_dtype(dtype)))

test/test_dtype_alu.py: 131 warnings
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:98: RuntimeWarning: invalid value encountered in cast
    numpy_value = op2[1](op1[1](an, bn).astype(_to_np_dtype(d2)), cn)

test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:98: RuntimeWarning: invalid value encountered in add
    numpy_value = op2[1](op1[1](an, bn).astype(_to_np_dtype(d2)), cn)

test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:98: RuntimeWarning: invalid value encountered in multiply
    numpy_value = op2[1](op1[1](an, bn).astype(_to_np_dtype(d2)), cn)

test/test_dtype_alu.py::TestDTypeALU::test_int32_midcast_float
test/test_dtype_alu.py::TestDTypeALU::test_int32_midcast_float
  /home/uuuvn/src/tinygrad/test/test_dtype_alu.py:98: RuntimeWarning: overflow encountered in multiply
    numpy_value = op2[1](op1[1](an, bn).astype(_to_np_dtype(d2)), cn)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================================== slowest 20 durations ===============================================================================================
52.04s call     test/test_ops.py::TestOps::test_broadcast_partial
23.27s call     test/test_dtype.py::TestAutoCastType::test_int_to_float_unary_func
23.10s call     test/test_ops.py::TestOps::test_broadcast_full
8.74s call     test/imported/test_indexing.py::TestIndexing::test_advancedindex
7.23s call     test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32
6.23s call     test/test_ops.py::TestOps::test_tan
5.28s call     test/test_dtype.py::TestInt8DType::test_upcast_ops
4.80s call     test/test_ops.py::TestOps::test_pow_full
4.76s call     test/test_dtype.py::TestUint8DType::test_upcast_ops
4.62s call     test/test_linearizer.py::TestHandCodedOpts::test_masked_upcast_wino_full
4.33s call     test/test_linearizer.py::TestKernelOpts::test_tensor_core_opts
3.73s call     test/test_dtype.py::TestDoubleDType::test_upcast_to_ops
3.72s call     test/test_ops.py::TestOps::test_pow
3.61s call     test/test_ops.py::TestOps::test_sin
3.50s call     test/test_dtype_alu.py::TestDTypeALU::test_float16_unary
3.47s call     test/test_dtype.py::TestInt64DType::test_upcast_to_ops
3.37s call     test/test_ops.py::TestOps::test_cos
3.15s call     test/test_dtype.py::TestInt32DType::test_upcast_to_ops
3.10s call     test/test_dtype.py::TestUint64DType::test_upcast_to_ops
2.97s call     test/test_dtype.py::TestTypeSpec::test_full
============================================================================================= short test summary info ==============================================================================================
FAILED test/test_dtype_alu.py::TestDTypeALU::test_float_midcast_int32 - AssertionError:
======================================================================== 1 failed, 596 passed, 72 skipped, 289 warnings in 93.84s (0:01:33) ========================================================================
```

All except one succeeded. remu issues

hikettei commented:
thanks so much!! I don't have an actual AMD machine to test it on, but after trying it under various conditions, it seems that the comparison for `x==inf` isn't working properly maybe?

hikettei commented:
https://github.com/tinygrad/tinygrad/pull/5185 needs to be merged to pass LLVM tests

hikettei commented:
After merging upstream, it started timing out with clang/metal. I'm giving up because there are too many compiler-related issues that I cannot handle.

hikettei commented:
Thank you for clarifying the bounty mechanism I was worried about! 
The theory behind the math should be right (including AMD, CLANG, and METAL), so the remaining issues are completely related to the compiler.

The reasons for the currently failing tests can be broken down as follows:

- [ ] Linter
    - It seems a bit difficult to keep the linter `fastmath.py` within about 150 lines.
    - but I want to refactor the code before merging.
- [ ] CLANG/METAL
    - they are timing out after merging a commit at some point.
    - All tests related to accuracy are passing.
    - I ran the CLANG tests in the local envs. It appears that running all tests causes a segmentation fault in `test/test_linearizer_failures.py::TestLinearizerFailures::test_failure_22`. When run individually, it works fine.
- [ ] AMD
    - REMU needs to be updated. (after other checkmarks become green.)

hikettei commented:
https://github.com/hikettei/tinygrad/blob/08633ea36600f9b6a5072835eb864f81856a3b7a/tinygrad/fastmath.py#L138-L142
we need to implement O(1) hash table algorithm by combining UOps to solve timing-out issues (currently O(n)). Currently, when scheduling payne_hanek(x), the combinations explode.

Qazalin commented:
cool, happy to help make progress on https://github.com/tinygrad/tinygrad/pull/5172 and these timeouts.
I'm incrementally supporting more of this branch's new things to remu https://github.com/Qazalin/remu/commit/eb31fa4f0e0021406e6a4d3e88c60ad6f02eb210.

hikettei commented:
Thank you for updating remu! but it looks like there are still some ops that have not been addressed?...

Qazalin commented:
yeah remu is still a work in progress.

hikettei commented:
I got that. I will work on the time-out issue in the meantime!

hikettei commented:
True. Thank you for clarifying and helping me!

`> you can help accelerate the fix with small isolated tests`

OK, I will push it

hikettei commented:
Reproduced the graph timeout:

```python
$ python -m pytest ./test/test_jit.py::TestJit::test_approx_sin_jit
```

chenyuxyz commented on test/test_const_folding.py:
```diff
@@ -14,7 +14,7 @@ def _check_ast_count(desired_count:int, t:Tensor):
 
 class TestUnaryOpsConstFolding(unittest.TestCase):
```
These tests should be updated properly, instead of just bumping the kernel counts. Positive counts means they are not folded. I think it's either keep as 0 and fixed the test, or keep as 0 and add expectedfailure (for the failed ones only and keep folded ones tested)

hikettei commented:
Thank you Qazalin for dealing with multi-tensor dtype issues! 

I have improved the time-out issue by simplifying the payne_hanek algorithm, and now the METAL tests can somehow be completed on CI. (CLANG still takes a longer bit >= 20min in `test/models/test_real_world.py::TestRealWorld::test_stable_diffusion`)

When `@TinyJIT` is enabled, `sin(x, fast=False)` does not compile within a reasonable time frame (e.g.: takes 840s with stable diffusion), and I would like to improve this. Especially, the computational load explodes when composing sin functions.

Currently, I'm running the test against:

```sh
# Set METAL=1 or CLANG=1
$ python -m pytest ./test/test_jit.py::TestJit::test_approx_jit_timeout -vv
```

Qazalin commented on tinygrad/function.py:
```diff
@@ -36,13 +37,19 @@ def forward(self, x:LazyBuffer) -> LazyBuffer:
 
-  def backward(self, grad_output:LazyBuffer) -> LazyBuffer:
-    return self.x.const(math.pi / 2).e(BinaryOps.ADD, self.x.e(UnaryOps.NEG)).e(UnaryOps.SIN).e(BinaryOps.MUL, grad_output)
+  def backward(self, grad_output: LazyBuffer) -> LazyBuffer:
```
Can you remove the space between `param` and `Type`? This is what we follow across the codebase:

```diff
- def backward(self, grad_output: LazyBuffer) -> LazyBuffer:
+ def backward(self, grad_output:LazyBuffer) -> LazyBuffer:
```

chenyuxyz reviewed (commented):
cool! i think this very close. we will discuss it in the upcoming Monday meeting (9:40 pt) to get it merged.

can you add a test to confirm these are fused into one kernel? similar to tests in test_schedule.py, and feel free to make a standalone test_fastmath.py

also can you add some reference / comment / link to the algorithms, this would make maintaining these easier in the future.

hikettei commented:
Thanks! I could not have reached this point without the support of community.

> can you add a test to confirm these are fused into one kernel? similar to tests in test_schedule.py, and feel free to make a standalone test_fastmath.py

- [ ] Sounds good, I'll have it written by tomorrow.

> also can you add some reference

I'll be happy to summarize my prep research.

-  Original Paper (SLEEF) https://arxiv.org/abs/2001.09258
- **Mock code in Julia.** (slightly older, but the main points are the same) https://github.com/tinygrad/tinygrad/pull/5085#issuecomment-2185557902
    - the algorithms do not perfectly match in some points, but they should help you to understand my codes.
- **Speed Improvements** https://github.com/tinygrad/tinygrad/pull/5085#issuecomment-2193801823
    - exp/log is the equivalent to the latest commit.
    - sin is **out of date** (but should be much faster in the latest commits. I will see it tomorrow)
   
- References for trig functions.
    - Implementing the reduction for sine was the most difficult part, so I have a ton of materials on it.
    - Reduction algorithm https://redirect.cs.umbc.edu/~phatak/645/supl/Ng-ArgReduction.pdf
    - Payne-Hanek in Julia https://gist.github.com/simonbyrne/d640ac1c1db3e1774cf2d405049beef3
    - **(Quite useful)** Payne-Hanek in C https://stackoverflow.com/questions/30463616/payne-hanek-algorithm-implementation-in-c
    - Payne-Hanek in CUDA https://forums.developer.nvidia.com/t/a-faster-and-more-accurate-implementation-of-sincosf/44620
    - The equivalent code in SLEEF https://github.com/shibatch/sleef/blob/4e08851f59fc2b545f9c393c6a23dfd311a26308/src/libm/sleefsp.c#L464

- [ ] TODO: write an explanation of the algorithm.

hikettei commented:
I have tweaked the fp16 sin by adding this line:

https://github.com/hikettei/tinygrad/blob/7534a0106d421a8728347a806e0d2c49c3cd981b/tinygrad/fastmath.py#L107-L108

after that, remu seems to encounter new unsupported instructions.

geohot commented:
Looking good! A few things to make it final:

1. Rename "fastmath.py" to "transcendental.py" https://en.wikipedia.org/wiki/Transcendental_function
2. This should only run by default for LLVM and CLANG, and the gate should be in function.py. TRANSCENDENTAL=0 should never run it and TRANSCENDENTAL=2 should always run it. (this way, you don't have to worry about remu)
3. Needs tests that it folds into a single kernel. #4492 might just have them.
4. We plan to raise the line count to 8500, you can put that in this PR to make CI green.

hikettei commented:
| Runtime | Op | Dtype | Base(m) | Approx(ms) |
|---|---|---|---|---|
|CLANG| sin | fp16 | 359.54 | 1687.48 | 
|CLANG| sin | fp32 | 1236.00 | 1413.16 |
|CLANG| sin | fp64 | 1221.10 | 2344.80 |
|CLANG| exp2 | fp16 | 187.13 | 177.90 |
|CLANG| exp2 | fp32 | 254.43 | 236.65 |
|CLANG| exp2 | fp64 | 528.48 | 650.17 |
|CLANG| log2 | fp16 | 233.28 | 235.58 |
|CLANG| log2 | fp32 | 304.36 | 268.34 |
|CLANG| log2 | fp64 | 911.42 | 861.83 |
|LLVM| sin | fp16 | 422.21 | 6529.47 |
|LLVM| sin | fp32 | 1301.17 | 1529.14 |
|LLVM| sin | fp64 | 1809.20 | 2376.0 |
|LLVM| exp2 | fp16 | 239.64 | 1743.73 |
|LLVM| exp2 | fp32 | 243.49 | 474.41 |
|LLVM| exp2 | fp64 | 487.33 | 761.60 |
|LLVM| log2 | fp16 | 295.27 | 300.21 |
|LLVM| log2 | fp32 | 281.34 | 552.40 |
|LLVM| log2 | fp64 | 461.11 | 718.07 |

hikettei commented:
1 . Moved to `tinygrad/transcendental.py`
2. Implemented. `TRANSCENDENTAL=0` to always ignore, `TRANSCENDENTAL=1` to only run in CLANG/LLVM (default), `TRANSCENDENTAL=2` to always run.
3. Added to `test/test_transcendental.py`
4. I have a benchmark (please refer to the table above). all ops except for LLVM+FP16+sin satisfy the speed requirement.

hikettei commented:
- Setting fast=True makes fp16+CLANG sin 10x times faster according to previous benchmarks. (with no accuracy drop in abs(x)<=9000)
    - a common strategy for fast sin approximation is to use different reduction algorithms depending on the range of input. However, we are running all reductions now. we are not getting any speed gains. 
- LLVM+FP16 is always slower than built_in. I will look into it.

geohot commented on tinygrad/function.py:
```diff
@@ -1,12 +1,24 @@
+  # TRANSCENDENTAL=0 to always ignore.
+  # TRANSCENDENTAL=1 to run only in CLANG/LLVM (default).
+  # TRANSCENDENTAL=2 to always run it.
+  lv = getenv("TRANSCENDENTAL", 1)
```
Should be a ContextVar in helpers so you can do "with(TRANSCENDENTAL=2):" 

Useful in the single kernel tests.

nimlgen replied:
@hikettei I see you added a ContextVar, but use it here as well `from tinygrad.helpers import TRANSCENDENTAL`

hikettei commented:
aah i see
For LLVM+FP16, i took the benchmark to bfloat16 instead of float16, the speed drop is reduced up to 1.5x times slower.

hikettei commented:
its ready

Qazalin commented on test/test_jit.py:
```diff
@@ -18,6 +20,18 @@ def _simple_test(add, extract=lambda x: x, N=10):
+  @settings(deadline=2e4)
+  @unittest.skipUnless(Device.DEFAULT in ["LLVM", "CLANG"], f"no support on {Device.DEFAULT}")
+  @given(strat.sampled_from([Tensor.exp2, Tensor.log2, Tensor.sin]))
+  def test_approx_jit_timeout(self, op):
```
wanna add the context var here too?

Qazalin commented on tinygrad/function.py:
```diff
@@ -1,12 +1,24 @@
+  # TRANSCENDENTAL=0 to always ignore.
+  # TRANSCENDENTAL=1 to run only in CLANG/LLVM (default).
+  # TRANSCENDENTAL=2 to always run it.
+  if TRANSCENDENTAL >= 2: return True
```
shouldn't it always check the dtype?

chenyuxyz commented:
approximation does not support bfloat16 right? why is there a speed difference in the table

hikettei commented:
Yes, as of now there's no bfloat16 support.
When benchmarking, I set `dtypes.default_float=dtypes.bfloat16` , but when calling `xexp(d)`, I found that the input was cast to `dtypes.double` now (why?).

hikettei commented:
Do you think there are more decent ways to benchmark? (I have excluded the compilation time.)

<details>
<summary>code</summary>

```python
# usage: LLVM=1 python benchmark.py
import time
from tinygrad import Tensor, dtypes, TinyJit
from tinygrad.helpers import Context

N = 100
K = 10000
DEVICE="LLVM"

def bench(op, transcendental, dtype):
  print(f"doing: {op}")
  with Context(TRANSCENDENTAL=transcendental):
     x = Tensor.ones(K, K, dtype=dtype).realize()
     o = op(x).realize()
     @TinyJit
     def f(x): return op(x).realize()
     if transcendental == 0: f = op
     # whole time
     t1 = time.perf_counter()
     for _ in range(N): o = f(x).numpy()
     t2 = time.perf_counter()
     return (t2 - t1) / N

def builtin_vs_approx(op_name:str, op, dtype_name:str, dtype):
    builtin = bench(op, 0, dtype)
    approx = bench(op, 2, dtype)
    return f"|{DEVICE}|{op_name}|{dtype_name}|{builtin}|{approx}|{approx / builtin}|"

def profile_device(op_name, op):
    return f"""{builtin_vs_approx(op_name, op, "float16", dtypes.float16)}
{builtin_vs_approx(op_name, op, "float32", dtypes.float32)}
{builtin_vs_approx(op_name, op, "float64", dtypes.float64)}"""

def profile_approx():
    return f"""{profile_device("sin", Tensor.sin)}
{profile_device("exp2", Tensor.exp2)}
{profile_device("log2", Tensor.log2)}"""

if __name__ == "__main__":
    print(profile_approx())

```
</details>

|DEVICE|op|dtype|built_in_time (ms)|approx_time (ms)| approx / built_in |
|---|---|---|---|---|---|
|LLVM|sin|float16|0.02583651209017262|0.024051261669956147|0.9309020345321487|
|LLVM|sin|float32|0.04163002958986908|0.04101716999895871|0.9852784252870308|
|LLVM|sin|float64|0.08432950334157795|0.08032227915944531|0.952481349665949|
|LLVM|exp2|float16|0.02530274583026767|0.031578660840168596|1.2480329625883348|
|LLVM|exp2|float32|0.042555369168985634|0.052152098331134766|1.2255115946483959|
|LLVM|exp2|float64|0.10032106374856084|0.08673374208854512|0.8645616269174514|
|LLVM|log2|float16|0.024750318750739096|0.03213025250006467|1.2981753012415322|
|LLVM|log2|float32|0.044224566670600325|0.04414197375066578|0.9981324199160678|
|LLVM|log2|float64|0.08472727458924055|0.09452266833046451|1.115610867795668|
|CLANG|sin|float16|0.027613637500908225|0.021477018331643193|0.7777685330640269|
|CLANG|sin|float32|0.04491972499992698|0.0394174079108052|0.8775077743879615|
|CLANG|sin|float64|0.08958673333982006|0.0815541870915331|0.9103377704618614|
|CLANG|exp2|float16|0.02527091542026028|0.02754374082898721|1.0899383884963958|
|CLANG|exp2|float32|0.05157496417174116|0.05261304209008813|1.0201275548130337|
|CLANG|exp2|float64|0.0993444904196076|0.0968433295795694|0.9748233562880649|
|CLANG|log2|float16|0.023417609168682248|0.027000463330186904|1.1529982901199076|
|CLANG|log2|float32|0.04541927500162274|0.05772871542023495|1.271017985605724|
|CLANG|log2|float64|0.08830231999978423|0.09300336375134066|1.0532380548049918|

geohot commented:
Will merge if tests pass, then bounty is yours!

After merge I'm going to move the gate to lazy.py instead of function.py. It makes more sense there
